aws_topology/stackstate_checks/aws_topology/resources/rds.py

```diff
@@ -1,89 +1,161 @@
+    VpcSecurityGroupId = StringType(required=True)
+
+
+class Instance(Model):
+    DBInstanceArn = StringType(required=True)
+    DBInstanceIdentifier = StringType(default="UNKNOWN")
+    DBSubnetGroup = ModelType(InstanceSubnetGroup)
+    Endpoint = ModelType(InstanceEndpoint)
+    VpcSecurityGroups = ListType(ModelType(InstanceVpcSecurityGroup))
 
 
 class RdsCollector(RegisteredResourceCollector):
     API = "rds"
     API_TYPE = "regional"
     COMPONENT_TYPE = "aws.rds_cluster"
 
-    def process_all(self, filter=None):
-        for instance_data_raw in self.client.describe_db_instances().get('DBInstances') or []:
-            instance_data = make_valid_data(instance_data_raw)
-            self.process_instance(instance_data)
+    @set_required_access_v2("rds:ListTagsForResource")
+    def collect_tags(self, resource_arn):
+        return self.client.list_tags_for_resource(ResourceName=resource_arn).get("TagList", "")
+
+    def collect_cluster(self, cluster_data):
+        tags = self.collect_tags(cluster_data.get("DBInstanceArn", "")) or []
+        return ClusterData(cluster=cluster_data, tags=tags)
+
+    def collect_clusters(self, **kwargs):
+        for cluster in [
+            self.collect_cluster(cluster_data)
+            for cluster_data in client_array_operation(self.client, "describe_db_clusters", "DBClusters", **kwargs)
+        ]:
+            yield cluster
+
+    def collect_instance(self, instance_data):
+        tags = tags = self.collect_tags(instance_data.get("DBClusterArn", "")) or []
+        return InstanceData(instance=instance_data, tags=tags)
 
-        for cluster_data_raw in self.client.describe_db_clusters().get('DBClusters') or []:
-            cluster_data = make_valid_data(cluster_data_raw)
+    def collect_instances(self, **kwargs):
+        for instance in [
+            self.collect_cluster(instance_data)
+            for instance_data in client_array_operation(self.client, "describe_db_instances", "DBInstances", **kwargs)
+        ]:
+            yield instance
+
+    @set_required_access_v2("rds:DescribeDBClusters")
+    def process_clusters(self):
+        for cluster_data in self.collect_clusters():
             self.process_cluster(cluster_data)
 
-    def process_one_cluster(self, id):
-        id = id.rsplit(':', 1)[-1]
-        for cluster_data_raw in self.client.describe_db_clusters(DBClusterIdentifier=id).get('DBClusters') or []:
-            cluster_data = make_valid_data(cluster_data_raw)
+    @set_required_access_v2("rds:DescribeDBInstances")
+    def process_instances(self):
+        for instance_data in self.collect_instances():
+            self.process_instance(instance_data)
+
+    def process_all(self, filter=None):
+        if not filter or "clusters" in filter:
+            self.process_clusters()
+        if not filter or "instances" in filter:
+            self.process_instances()
+
+    def process_one_cluster(self, cluster_arn):
+        cluster_id = cluster_arn.rsplit(":", 1)[-1]
```

Comment:
Guess this can become a "safe" function in utils that returns "" in case the is no colon. (Also add test for that utils)

Reply:
I did a test and found that the API accepts either the identifier or the full ARN, so now I just pass in the full ARN